mutation/index.ts

```diff
@@ -0,0 +1,132 @@
+import { useCallback, useRef } from 'react'
+import useSWR, {
+  useSWRConfig,
+  Middleware,
+  Key,
+  Fetcher,
+  SWRResponse
+} from 'swr'
+
+import { serialize } from '../src/utils/serialize'
+import { useStateWithDeps } from '../src/utils/state'
+import { withMiddleware } from '../src/utils/with-middleware'
+import { useIsomorphicLayoutEffect } from '../src/utils/env'
+import { isUndefined, UNDEFINED } from '../src/utils/helper'
+import { getTimestamp } from '../src/utils/timestamp'
+
+import { SWRMutationConfiguration, SWRMutationResponse } from './types'
+
+const mutation = <Data, Error>() => (
+  key: Key,
+  fetcher: Fetcher<Data>,
+  config: SWRMutationConfiguration<Data, Error> = {}
+) => {
+  const { mutate } = useSWRConfig()
+
+  const keyRef = useRef(key)
+  // Ditch all mutation results that happened earlier than this timestamp.
+  const ditchMutationsTilRef = useRef(0)
+
+  const [stateRef, stateDependencies, setState] = useStateWithDeps<Data, Error>(
+    {
+      data: UNDEFINED,
+      error: UNDEFINED,
+      isValidating: false
+    },
+    true
+  )
+  const currentState = stateRef.current
+
+  // Similar to the global mutate, but bound to the current cache and key.
+  // `cache` isn't allowed to change during the lifecycle.
+  const boundMutate: SWRResponse<Data, Error>['mutate'] = useCallback(
+    (newData, shouldRevalidate) => {
+      const serializedKey = serialize(keyRef.current)[0]
+      return mutate(serializedKey, newData, shouldRevalidate)
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    []
+  )
+
+  const trigger = useCallback(async (extraArg, opts) => {
+    if (!fetcher) {
+      throw new Error('Can’t trigger the mutation: missing fetcher.')
+    }
+
+    const [serializedKey, args] = serialize(keyRef.current)
+    const options = Object.assign({}, config, opts)
+
+    // Disable cache population by default.
+    if (isUndefined(options.populateCache)) {
+      options.populateCache = false
+    }
+
+    // Trigger a mutation, also track the timestamp. Any mutation that happened
+    // earlier this timestamp should be ignored.
+    const mutationStartedAt = getTimestamp()
+    ditchMutationsTilRef.current = mutationStartedAt
+
+    try {
```

Comment:
maybe move `setState` and `args.push` calls outside of try-catch scope? like you did in previous refactor PRs